server/push/tnpg/push_tnpg.go

```diff
@@ -245,14 +261,14 @@ func handlePushResponse(batch *batchResponse, messages []fcm.MessageData) {
 	}
 }
 
-func handleSubResponse(batch *batchResponse, req *push.ChannelReq) {
+func handleSubResponse(batch *batchResponse, req *push.ChannelReq, devices []string) {
 	if batch.FailureCount <= 0 {
 		return
 	}
 
 	for _, resp := range batch.Responses {
 		// FCM documentation sucks. There is no list of possible errors so no action can be taken but logging.
-		log.Println("fcm sub/unsub error", resp.ErrorCode, req.Uid, req.Devices[resp.Index])
+		log.Println("fcm sub/unsub error", resp.ErrorCode, req.Uid, devices[resp.Index])
```

Comment:
Check `resp.ErrorCode` to avoid excessive logging?

Reply:
The way I understand it, it works differently from pushes: it returns errors only for those records which produced an error. The `Index` is a reference to the original array of device IDs. In other words, `resp.ErrorCode` is always non-blank.